source/RoslynAnalyzers/CommandHandlerResponseConstructorParameterAnalyzer.cs

```diff
@@ -0,0 +1,100 @@
+        const string Description = "IResponse forms part of the boundary between the HTTP and API layers of the Octopus Server. Domain models should not leave the API, so implementations of IResponse should not take any model types.";
+
+        const string NotAllowedNamespace = "Octopus.Core.Model";
+
+        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            Title,
+            MessageFormat,
+            Category,
+            DiagnosticSeverity.Error,
+            true,
+            Description
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(DetectIfIResponseConstructorPassesAModelType, SyntaxKind.ClassDeclaration);
+        }
+
+        static void DetectIfIResponseConstructorPassesAModelType(SyntaxNodeAnalysisContext context)
+        {
+            if (
+                context.SemanticModel.GetDeclaredSymbol(context.Node) is INamedTypeSymbol symbol &&
+                ImplementsIResponse(symbol))
+            {
+                var constructors = symbol?.Constructors.ToList();
+                if (constructors != null)
+                {
+                    foreach (var constructor in constructors)
+                    {
+                        var ctorParameters = constructor.Parameters.ToList();
+                        foreach (var parameterSymbol in ctorParameters)
+                        {
+                            if (parameterSymbol != null &&
+                                parameterSymbol.Type.ContainingNamespace != null)
+                            {
+                                var constituentNamespaces = parameterSymbol.Type.ContainingNamespace.ConstituentNamespaces;
+                                if (constituentNamespaces != null)
+                                {
+                                    foreach (var namespaceSymbol in constituentNamespaces)
+                                    {
+                                        if (namespaceSymbol != null && namespaceSymbol.ToString().StartsWith(NotAllowedNamespace))
+                                        {
+                                            var parameterDeclaringSyntaxReference = parameterSymbol.DeclaringSyntaxReferences.FirstOrDefault();
+                                            if (parameterDeclaringSyntaxReference != null)
+                                            {
+                                                var diagnostic = Diagnostic.Create(Rule, parameterDeclaringSyntaxReference.GetSyntax().GetLocation());
+                                                context.ReportDiagnostic(diagnostic);
```

Comment:
This level of nesting is pretty hard to read. You should be able to convert this into a LINQ query which would be easier to understand.

Reply:
Paul and I discussed this earlier on - a refactor should be on its way soon.